pkg/fqdn/re/re.go

```diff
@@ -0,0 +1,55 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+// Package re provides a simple function to access compile regex objects for
+// the FQDN subsystem.
+package re
+
+import (
+	"fmt"
+	"regexp"
+
+	lru "github.com/golang/groupcache/lru"
+)
+
+// CompileRegex provides a conviennce wrapper around the LRU. It will either
+// return the cached regex object or compile the pattern into a regex object if
+// the LRU hasn't been initialized or it doesn't exist in the cache. If it
+// doesn't exist in the cache, it will be added.
+func CompileRegex(p string) (*regexp.Regexp, error) {
+	if RegexCompileLRU == nil {
+		return regexp.Compile(p)
```

Comment:
This is the area that I kinda struggled to come to a conclusion on. What I opted for here at a high-level is to ignore that initialization didn't happen and simply pass-through straight into a regex compile, transparent to the caller. And that was the biggest motivation--to be transparent. But with that comes that cost of hiding when a human error was made.

I just realized while typing this out that we could have a sync.Once inside this case to initialize the cache if it wasn't, so that should catch any "premature" calls to `CompileRegex` that were done without initializing the LRU first (in case of human error or even otherwise).

I'll implement that, thanks.

Reply:
Hm, the initialization function needs to take an int for the size which we don't necessarily have with CompileRegex and would be awkward to always pass the size of the LRU to it... :thinking:

Reply:
Fixed by using option.Config.FQDNRegexCompileLRUSize. If the init fails, we'll log a warning. I don't expect to hit this case going forward as the initialization happens during the DNS proxy bootstrap. Assuming the shuffling of subsystem inits don't occur, this should never happen. If it does happen, we'll have a warning log. LMK if you think this addresses the concern or if we should think about a deeper approach to the problem.

Reply:
Minor nit here, this now initializes the LRU on the first pass, but if the initialization is successful it will actually not cache that one first compiled regex. That said, this is also not such a big deal since if the same regex gets used many times, then it will end up in the LRU soon anyhow.

Reply:
Hmm. Multiple paths for initialization is a bit awkward for some of the other reasons I've just posted. I think that this ends up making the code more complicated in a sort of "defensive coding" kind of way. If we can just guarantee that the structure is initalized early enough, then we don't need to worry about any of this extra logic.

Reply:
We can guarantee that the DNS proxy in a non-test context will always initialize the LRU before serving DNS responses because we've put the LRU init inside the DNS proxy init.

What we can't really guarantee is if the daemon init ordering is for some reason shuffled around such that the K8s resources start syncing before the DNS proxy is up (and LRU is init'd). In that case, the LRU will not be initialized at first so we could be compiling regexes without caching them for a period of time, until the DNS proxy init starts. 

I'm not sure there is any way around this, at least until we have refectored the daemon init such that we can explicitly control which steps must happen before others, and so on.

This is why the original code was checking if the LRU was nil and just passing straight down to the regex compile call without doing anything else, because we assume that eventually, as long as we aren't in the testing context, that the LRU will get init'd.

Hopefully that clarifies this issue. So if the defensive coding is not really worth it, happy to remove it and go back to the original code as mentioned above.

Reply:
> What we can't really guarantee is if the daemon init ordering is for some reason shuffled around such that the K8s resources start syncing before the DNS proxy is up (and LRU is init'd). In that case, the LRU will not be initialized at first so we could be compiling regexes without caching them for a period of time, until the DNS proxy init starts.

I was fishing for is a solution to this. If we don't provide this guarantee, then we could be spending more time/effort/memory on these regex compilations. For that undetermined period on startup, the cache does not work at all. Worst case, it makes the cache completely ineffective on startup and then it's only useful over time when new policies are added. My hope with the "error" suggestion would be that if the initialization returns an error up the stack, then at some point this will either trigger an agent-wide `log.Fatal()` or some sort of prominent error message to be logged. Then we catch this as part of testing in CI, detect whenever the guarantee is no longer provided, and make sure we fix it before that code gets into a release. That said, I'm not attached to this solution.

In my last comment, I was thinking that ideally there's just one place where we guarantee that the LRU is initialized. If we're going to make this Compile function more complicated to handle this case, then is it still useful to explicitly initialize the LRU from the DNS proxy? Presumably the first time that the DNS proxy uses this code, it will also call into the same Compile function. Is the DNS proxy initialization redundant, then? Maybe we can just contain the "worry about LRU initalization" logic to this Compile function?

(The other alternative is just explicitly initialize it super early even directly from the daemon init routines or something, but I'm not overly sold on that approach either).

Reply:
Ok fair enough. Apologies this took longer than it should have. Yes returning an error will just simplify things. 

> My hope with the "error" suggestion would be that if the initialization returns an error up the stack, then at some point this will either trigger an agent-wide log.Fatal() or some sort of prominent error message to be logged. Then we catch this as part of testing in CI, detect whenever the guarantee is no longer provided, and make sure we fix it before that code gets into a release. That said, I'm not attached to this solution.

This didn't sound great to me at first because it's kind of a "stop the world" error, but it's not a bad idea now that I think about it a bit more. We won't run into this often anyway, if it all.

Reply:
Implemented.

Reply:
:+1: No need to apologise, it's good practice to have fully discussion like this to try to figure out the ideal implementation or challenge our assumptions. You also can't read my mind unless I put it into words first :)